### ISSUE TITLE
SERVFAIL when CD set

### ISSUE THREAD

**Issue by FiloSottile** [getdnsapi/stubby#48]
```
$ dig @127.0.0.1 -p5342 filippo.io

; <<>> DiG 9.11.2 <<>> @127.0.0.1 -p5342 filippo.io
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 8756
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 2, ADDITIONAL: 5

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
; COOKIE: f57cae418af1e3c353d164c85a2027b18645e5d70e9a2187 (good)
; CLIENT-SUBNET: 0/0/0
;; QUESTION SECTION:
;filippo.io.			IN	A

;; ANSWER SECTION:
filippo.io.		300	IN	A	104.20.145.15
filippo.io.		300	IN	A	104.20.144.15

;; AUTHORITY SECTION:
filippo.io.		83174	IN	NS	jim.ns.cloudflare.com.
filippo.io.		83174	IN	NS	beth.ns.cloudflare.com.

;; ADDITIONAL SECTION:
jim.ns.cloudflare.com.	23151	IN	A	173.245.59.125
beth.ns.cloudflare.com.	40084	IN	A	173.245.58.103
jim.ns.cloudflare.com.	23151	IN	AAAA	2400:cb00:2049:1::adf5:3b7d
beth.ns.cloudflare.com.	40084	IN	AAAA	2400:cb00:2049:1::adf5:3a67

;; Query time: 105 msec
;; SERVER: 127.0.0.1#5342(127.0.0.1)
;; WHEN: Thu Nov 30 15:45:57 GMT 2017
;; MSG SIZE  rcvd: 392

$ dig @127.0.0.1 -p5342 filippo.io +cd

; <<>> DiG 9.11.2 <<>> @127.0.0.1 -p5342 filippo.io +cd
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 9973
;; flags: qr rd cd; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 0
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;filippo.io.			IN	A

;; Query time: 3002 msec
;; SERVER: 127.0.0.1#5342(127.0.0.1)
;; WHEN: Thu Nov 30 15:46:02 GMT 2017
;; MSG SIZE  rcvd: 28
```

My config:

```
resolution_type: GETDNS_RESOLUTION_STUB
dns_transport_list:
  - GETDNS_TRANSPORT_TLS
tls_authentication: GETDNS_AUTHENTICATION_REQUIRED
tls_query_padding_blocksize: 256
edns_client_subnet_private : 1
idle_timeout: 60000
listen_addresses:
  - 127.0.0.1@5342
round_robin_upstreams: 0
tls_backoff_time: 300
timeout: 3000
upstream_recursive_servers:
  - address_data: 145.100.185.15
    tls_auth_name: "dnsovertls.sinodun.com"
    tls_pubkey_pinset:
      - digest: "sha256"
        value: 62lKu9HsDVbyiPenApnc4sfmSYTHOVfFgL3pyB+cBL4=
  - address_data: 145.100.185.16
    tls_auth_name: "dnsovertls1.sinodun.com"
    tls_pubkey_pinset:
      - digest: "sha256"
        value: cE2ecALeE5B+urJhDrJlVFmf38cJLAvqekONvjvpqUA=
  - address_data: 185.49.141.37
    tls_auth_name: "getdnsapi.net"
    tls_pubkey_pinset:
      - digest: "sha256"
        value: foxZRnIh9gZpWnl+zEiKa0EJ2rdCGroMWm02gaxSc9Q=
  - address_data: 89.233.43.71
    tls_auth_name: "unicast.censurfridns.dk"
    tls_pubkey_pinset:
      - digest: "sha256"
        value: wikE3jYAA6jQmXYTr/rbHeEPmC78dQwZbQp6WdrseEs=
  - address_data: 51.15.70.167
    tls_auth_name: "dns1.darkmoon.is"
    tls_pubkey_pinset:
      - digest: "sha256"
        value: 8sx8niFUiJvMM3C1qLE9cH79TuQQztzMVDtbKjpD/IQ=
  - address_data: 146.185.167.43
    tls_auth_name: "securedns.eu"
    tls_pubkey_pinset:
      - digest: "sha256"
        value: 2EfbwDyk2zSnAbBJSpCSWZKKGUD+a6p/yg2bxdC+x2A=
  - address_data: 81.187.221.24
    tls_auth_name: "dns-tls.bitwiseshift.net"
    tls_pubkey_pinset:
      - digest: "sha256"
        value: YmcYWZU5dd2EoblZHNf1jTUPVS+uK3280YYCdz4l4wo=
  - address_data: 194.130.110.185
    tls_auth_name: "ns1.dnsprivacy.at"
    tls_pubkey_pinset:
      - digest: "sha256"
        value: vqVQ9TcoR9RDY3TpO0MTXw1YQLjF44zdN3/4PkLwtEY=
  - address_data: 94.130.110.178
    tls_auth_name: "ns2.dnsprivacy.at"
    tls_pubkey_pinset:
      - digest: "sha256"
        value: s5Em89o0kigwfBF1gcXWd8zlATSWVXsJ6ecZfmBDTKg=
```

Curiously, Stubby sends a SERVFAIL precisely after the timeout has elapsed, but no timeout is logged.

```
[15:45:36.881302] STUBBY: Read config from file /usr/local/etc/stubby/stubby.yml
[15:45:36.883408] STUBBY: Starting DAEMON....
[15:45:39.027004] STUBBY: 145.100.185.15                           : Conn opened: TLS - Strict Profile
[15:46:29.712322] STUBBY: 145.100.185.15                           : Conn closed: TLS - Resps=     7, Timeouts  =     0, Curr_auth =Success, Keepalive(ms)= 60000
[15:46:29.712369] STUBBY: 145.100.185.15                           : Upstream   : TLS - Resps=     7, Timeouts  =     0, Best_auth =Success
[15:46:29.712379] STUBBY: 145.100.185.15                           : Upstream   : TLS - Conns=     1, Conn_fails=     0, Conn_shuts=      1, Backoffs     =     0
```

This is stopping me from using unbound behind stubby for caching and validation.

**Comment by FiloSottile** [comment]
Patching this line into `if (0)` "fixes" it.

https://github.com/getdnsapi/stubby/blob/5e27718b5b9baa810edc1a690bae6600b0ebc72b/src/stubby.c#L517

**Comment by wtoorop** [comment]
Yes, the getdns_return_all_statuses extension does not disable DNSSEC, it simply return all DNSSEC statuses (so also BOGUS answers).  All DNSSEC queries will be made with that extension (causing timeouts sometimes, especially when the upstream does not support out of order processing of pipelined queries).  Unbound would in principle have the same issue when it would do the DNSSEC validation via Stubby, but is less likely to bump into it because of it's  caching.
So we need a small cache in getdns, if only just for DNSSEC answers, for better serviceability, but a short-term more easy fix would be to create another extension: "dnssec_checking_disabled", that would simply disable DNSSEC fetching and checking.  I'll try to have that in a shortly to do bugfix release

**Comment by wtoorop** [comment]
Released with getdns-1.3.0
